### PR TITLE
Bring your own class

### PR DESCRIPTION
**Pull request by piotrm0**
# Custom class Apps

This wrapper covers apps that are not based on one of the high-level frameworks
such as langchain or llama-index. We instead assume that some python class or
classes implements an app which has similar functionality to LLM apps coded in
the high-level frameworks in that it generally processes text queries to produce
text outputs while making intermediate queries to things like LLMs, vector DBs,
and similar.

## Example Usage

Consider a mock question-answering app with a context retriever component coded
up as two classes in two python, `CustomApp` and `CustomRetriever`:

### `custom_app.py`

```python
from trulens_eval.tru_custom_app import instrument
from custom_retriever import CustomRetriever 


class CustomApp:
    # NOTE: No restriction on this class.

    def __init__(self):
        self.retriever = CustomRetriever()

    @instrument
    def retrieve_chunks(self, data):
        return self.retriever.retrieve_chunks(data)

    @instrument
    def respond_to_query(self, input):
        chunks = self.retrieve_chunks(input) output = f"The answer to {input} is
        probably {chunks[0]} or something ..." return output
```

### `custom_retriever.py`

```python
from trulens_eval.tru_custom_app import instrument

class CustomRetriever:
    # NOTE: No restriction on this class either.

    @instrument
    def retrieve_chunks(self, data):
        return [
            f"Relevant chunk: {data.upper()}", f"Relevant chunk: {data[::-1]}"
        ]
```

The core tool for instrumenting these classes is the `instrument` method
(actually class, but details are not important here). trulens needs to be aware
of two high-level concepts to usefully monitor the app: components and methods
used by components. The `instrument` must decorate each method that the user
wishes to watch (for it to show up on the dashboard). In the example, all of the
functionalities are decorated. Additionally, the owner classes of any decorated
method is viewed as an app component. In this case `CustomApp` and
`CustomRetriever` are components. 

Following the instrumentation, the app can be used with or without tracking:

### `example.py`

```python
from custom_app import CustomApp from trulens_eval.tru_custom_app
import TruCustomApp

ca = CustomApp()

# Normal app usage:
response = ca.respond_to_query("What is the capital of Indonesia?")

# Wrapping app with `TruCustomApp`: 
ta = TruCustomApp(ca)

# Wrapped usage: must use the general `with_record` (or `awith_record`) method:
response, record = ta.with_record(
    ca.respond_to_query, input="What is the capital of Indonesia?"
)
```

The `with_record` use above returns both the response of the app normally
produces as well as the record of the app as is the case with the higher-level
wrappers. `TruCustomApp` constructor arguments are like in those higher-level
apps as well including the feedback functions, metadata, etc.

### Instrumenting 3rd party classes

In cases you do not have access to a class to make the necessary decorations for
tracking, you can instead use one of the static methods of `instrument`, for
example, the alterative for making sure the custom retriever gets instrumented
is via:

```python
# custom_app.py`:

from trulens_eval.tru_custom_app import instrument
from somepackage.from custom_retriever import CustomRetriever

instrument.method(CustomRetriever, "retrieve_chunks")

# ... rest of the custom class follows ...
```

## API Usage Tracking

Uses of python libraries for common LLMs like OpenAI are tracked in custom class
apps.

### Covered LLM Libraries

- Official OpenAI python package (https://github.com/openai/openai-python).

### Huggingface

Uses of huggingface inference APIs are tracked as long as requests are made
through the `requests` class's `post` method to the URL
https://api-inference.huggingface.co .

## Limitations

- Tracked (instrumented) components must be accessible through other tracked
  components. Specifically, an app cannot have a custom class that is not
  instrumented but that contains an instrumented class. The inner instrumented
  class will not be found by trulens.

- All tracked components are categorized as "Custom" (as opposed to Template,
  LLM, etc.). That is, there is no categorization available for custom
  components. They will all show up as "uncategorized" in the dashboard.

- Non json-like contents of components (that themselves are not components) are
  not recorded or available in dashboard. This can be alleviated to some extent
  with the `app_extra_json` argument to `TruCustomClass` as it allows one to
  specify in the form of json additional information to store alongside the
  component hierarchy. Json-like (json bases like string, int, and containers
  like sequences and dicts are included).

## What can go wrong

- If a `with_record` or `awith_record` call does not encounter any instrumented
  method, it will raise an error. You can check which methods are instrumented
  using `App.print_instrumented`. You may have forgotten to decorate relevant
  methods with `@instrument`.

```python
app.print_instrumented()

### output example:
Components:
Custom of trulens_eval.app component: *.__app__.app
Custom of trulens_eval.app component: *.__app__.app.llm
Custom of trulens_eval.app component: *.__app__.app.retriever
Custom of trulens_eval.app component: *.__app__.app.template

Methods:
Object at 0x1064c6d30:
	<function CustomApp.retrieve_chunks at 0x14c6c5700> with path *.__app__.app
	<function CustomApp.respond_to_query at 0x14c6c5790> with path *.__app__.app
Object at 0x1064c6dc0:
	<function CustomLLM.generate at 0x14c6c51f0> with path *.__app__.app.llm
Object at 0x1064c6f70:
	<function CustomRetriever.retrieve_chunks at 0x14c6b5c10> with path *.__app__.app.retriever
Object at 0x106272100:
	<function CustomTemplate.fill at 0x14c6c54c0> with path *.__app__.app.template
```

- If an instrumented / decorated method's owner object cannot be found when
  traversing your custom class, you will get a warning. This may be ok in the
  end but may be indicative of a problem. Specifically, note the "Tracked"
  limitation above. You can also use the `app_extra_json` argument to `App` /
  `TruCustomApp` to provide a structure to stand in place for (or augment) the
  data produced by walking over instrumented components to make sure this
  hierarchy contains the owner of each instrumented method.

  The owner-not-found error looks like this:

```python
Function <function CustomRetriever.retrieve_chunks at 0x177935d30> was not found during instrumentation walk. Make sure it is accessible by traversing app <custom_app.CustomApp object at 0x112a005b0> or provide a bound method for it as TruCustomApp constructor argument `methods_to_instrument`.
Function <function CustomTemplate.fill at 0x1779474c0> was not found during instrumentation walk. Make sure it is accessible by traversing app <custom_app.CustomApp object at 0x112a005b0> or provide a bound method for it as TruCustomApp constructor argument `methods_to_instrument`.
Function <function CustomLLM.generate at 0x1779471f0> was not found during instrumentation walk. Make sure it is accessible by traversing app <custom_app.CustomApp object at 0x112a005b0> or provide a bound method for it as TruCustomApp constructor argument `methods_to_instrument`.
```

  Subsequent attempts at `with_record`/`awith_record` may result in the "Empty
  record" exception.

- Usage tracking not tracking. We presently have limited coverage over which
  APIs we track and make some assumptions with regards to accessible APIs
  through lower-level interfaces. Specifically, we only instrument the
  `requests` module's `post` method for the lower level tracking. Please file an
  issue on github with your use cases so we can work out a more complete
  solution as needed.

### CODE REVIEW

File: trulens_eval/trulens_eval/tru_custom_app.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+        self.func = func
+
+    def __set_name__(self, cls, name):
+        # Add owner of the decorated method, its module, and the name to the
+        # Default instrumentation walk filters.
+        Instrument.Default.MODULES.add(cls.__module__)
+        Instrument.Default.CLASSES.add(cls)
+        Instrument.Default.METHODS[
+            name] = lambda o: True  # lambda o: isinstance(o, cls)
+        # TODO: fix the last line in case a method with the same name appears in
```
Comment (lariel-fernandes) [inline trulens_eval/trulens_eval/tru_custom_app.py:304]: A possible fix:
```python
check_o = Instrument.Default.METHODS.get(name, lambda o: False)
Instrument.Default.METHODS[name] = lambda o: check_o(o) or isinstance(o, cls)
```

**Comment by lariel-fernandes** [comment]
Sometimes the custom App that I want to instrument may be defined in a package that I don't control, so I would need to subclass it just to apply the decorator:
```python
from package import CustomApp

class InstrumentedApp(CustomApp):
    @instrument
    def method(self, **kwargs):
        return super().method(**kwargs)
```
To make that less convoluted we could extract the code from `instrument.__set_name__` into a separate function, which can be called on its own:
```python
from package import CustomApp

instrument_methods(CustomApp, method_names=["some_method", "other_method"])
```
Then in `instrument.__set_name__` we just need to call that function too.

**Comment by piotrm0** [comment]
> Sometimes the custom App that I want to instrument may be defined in a package that I don't control, so I would need to subclass it just to apply the decorator:
> 
> ```python
> from package import CustomApp
> 
> class InstrumentedApp(CustomApp):
>     @instrument
>     def method(self, **kwargs):
>         return super().method(**kwargs)
> ```
> 
> To make that less convoluted we could extract the code from `instrument.__set_name__` into a separate function, which can be called on its own:
> 
> ```python
> from package import CustomApp
> 
> instrument_methods(CustomApp, method_names=["some_method", "other_method"])
> ```
> 
> Then in `instrument.__set_name__` we just need to call that function too.

Good idea; I added this as `instrument.method` and `instrument.methods`.

**Review by shayaks — APPROVED** [review]

File: trulens_eval/examples/frameworks/custom/custom_example.ipynb
[the remark this answers is not in the document]
Comment (piotrm0) [reply trulens_eval/examples/frameworks/custom/custom_example.ipynb:1]: This example is based on several files which might not work in colab.

**Pull request merged**